autoPyTorch/data/utils.py

```diff
@@ -0,0 +1,309 @@
+        if set(dataset_compression.keys()) != set(default_dataset_compression_arg.keys()):
+            raise ValueError(
+                f"Unknown key in dataset_compression, {list(dataset_compression.keys())}."
+                f"\nPossible keys are {list(default_dataset_compression_arg.keys())}"
+            )
+
+        memory_allocation = dataset_compression["memory_allocation"]
+
+        # "memory_allocation" must be float or int
+        if not (isinstance(memory_allocation, float) or isinstance(memory_allocation, int)):
+            raise ValueError(
+                "key 'memory_allocation' must be an `int` or `float`"
+                f"\ntype = {memory_allocation}"
+                f"\ndataset_compression = {dataset_compression}"
+            )
+
+        # "memory_allocation" if absolute, should be > 0 and < memory_limit
+        if isinstance(memory_allocation, int) and not (0 < memory_allocation < memory_limit):
+            raise ValueError(
+                f"key 'memory_allocation' if int must be in (0, memory_limit={memory_limit})"
+                f"\nmemory_allocation = {memory_allocation}"
+                f"\ndataset_compression = {dataset_compression}"
+            )
+
+        # "memory_allocation" must be in (0,1) if float
+        if isinstance(memory_allocation, float):
+            if not (0.0 < memory_allocation < 1.0):
+                raise ValueError(
+                    "key 'memory_allocation' if float must be in (0, 1)"
+                    f"\nmemory_allocation = {memory_allocation}"
+                    f"\ndataset_compression = {dataset_compression}"
+                )
+            # convert to int so we can directly use
+            dataset_compression["memory_allocation"] = floor(memory_allocation * memory_limit)
+
+        # "methods" must be non-empty sequence
+        if (
+            not isinstance(dataset_compression["methods"], Sequence)
+            or len(dataset_compression["methods"]) <= 0
+        ):
+            raise ValueError(
+                "key 'methods' must be a non-empty list"
+                f"\nmethods = {dataset_compression['methods']}"
+                f"\ndataset_compression = {dataset_compression}"
+            )
+
+        # "methods" must contain known methods
+        if any(
+            method not in cast(Sequence, default_dataset_compression_arg["methods"])  # mypy
+            for method in dataset_compression["methods"]
+        ):
+            raise ValueError(
+                f"key 'methods' can only contain {default_dataset_compression_arg['methods']}"
+                f"\nmethods = {dataset_compression['methods']}"
+                f"\ndataset_compression = {dataset_compression}"
+            )
+
+        return cast(DatasetCompressionSpec, dataset_compression)
+    else:
+        raise ValueError(
+            f"Unknown type for `dataset_compression` {type(dataset_compression)}"
+            f"\ndataset_compression = {dataset_compression}"
+        )
```

Comment:
```suggestion
    if not isinstance(dataset_compression, Mapping):
        raise ValueError(
            f"Unknown type for `dataset_compression` {type(dataset_compression)}"
            f"\ndataset_compression = {dataset_compression}"
        )

    # Fill with defaults if they don't exist
    dataset_compression = {
        **default_dataset_compression_arg,
        **dataset_compression
    }

    # Must contain known keys
    if set(dataset_compression.keys()) != set(default_dataset_compression_arg.keys()):
        raise ValueError(
            f"Unknown key in dataset_compression, {list(dataset_compression.keys())}."
            f"\nPossible keys are {list(default_dataset_compression_arg.keys())}"
        )

    memory_allocation = validate_memory_allocation(dataset_compression["memory_allocation"], memory_limit)

    # "methods" must be non-empty sequence
    if (
        not isinstance(dataset_compression["methods"], Sequence)
        or len(dataset_compression["methods"]) <= 0
    ):
        raise ValueError(
            "key 'methods' must be a non-empty list"
            f"\nmethods = {dataset_compression['methods']}"
            f"\ndataset_compression = {dataset_compression}"
        )

    # "methods" must contain known methods
    if any(
        method not in cast(Sequence, default_dataset_compression_arg["methods"])  # mypy
        for method in dataset_compression["methods"]
    ):
        raise ValueError(
            f"key 'methods' can only contain {default_dataset_compression_arg['methods']}"
            f"\nmethods = {dataset_compression['methods']}"
            f"\ndataset_compression = {dataset_compression}"
        )

    return cast(DatasetCompressionSpec, dataset_compression)


def validate_memory_allocation(memory_allocation: Union[int, float], memory_limit: int) -> float:
    if not (isinstance(memory_allocation, float) or isinstance(memory_allocation, int)):
        raise ValueError(
            "'memory_allocation' must be an `int` or `float`"
            f"\nmemory_allocation = {memory_allocation}"
        )

    # "memory_allocation" if absolute, should be > 0 and < memory_limit
    if isinstance(memory_allocation, int) and not (0 < memory_allocation < memory_limit):
        raise ValueError(
            f"'memory_allocation' if int must be in (0, memory_limit={memory_limit})"
            f"\nmemory_allocation = {memory_allocation}"
        )

    if isinstance(memory_allocation, float):
        if not (0.0 < memory_allocation < 1.0):
            raise ValueError(
                "'memory_allocation' if float must be in (0, 1)"
                f"\nmemory_allocation = {memory_allocation}"
            )
        # convert to int so we can directly use
        memory_allocation = floor(memory_allocation * memory_limit)
    
    return memory_allocation
```

Reply:
I dont think this is necessary.

Reply:
Then could you at least perform beforehand so that we can reduce the nest?:
```python
    if not isinstance(dataset_compression, Mapping):
        raise ValueError(
            f"Unknown type for `dataset_compression` {type(dataset_compression)}"
            f"\ndataset_compression = {dataset_compression}"
        )
```

Reply:
yeah sure, that makes sense.